### PR TITLE
Timeline UI Tests

### PR DESCRIPTION
**Pull request by pixlwave**
This PR adds the main remaining tests for #352.

It also introduces `UITestSignalServer`/`Client` as a way of communicating across processes using UDP. In the future these could be changed to `TCP` and expanded upon, but for now there's a simple enough implementation to send and await a single signal at a time.

Note: The changelog already exists from a previous PR.

### CODE REVIEW

File: UITests/Sources/RoomScreenUITests.swift [hunk 1]
```diff
@@ -47,19 +47,123 @@ class RoomScreenUITests: XCTestCase {
...
+        let signal = try UITestSignalServer()
+        
         let app = Application.launch()
         app.goToScreenWithIdentifier(.roomSmallTimelineIncomingAndSmallPagination)
         
-        // Wait for both the incoming message and the pagination chunk.
-        XCTAssert(app.staticTexts["Bob"].waitForExistence(timeout: 2))
-        XCTAssert(app.staticTexts["Helena"].waitForExistence(timeout: 2))
+        // When a back pagination occurs and an incoming message arrives.
+        try await signal.connect()
```
Comment (stefanceriu) [inline UITests/Sources/RoomScreenUITests.swift:57]: Feels weird to have `signal` `connect` to anything. We should probably use a different name e.g. plainly `server`

File: ElementX/Sources/UITests/UITestSignalling.swift [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    case awaitingAnotherSignal
+    /// A network error occurred.
+    case nwError(NWError)
+    /// An unexpected signal was received. This error isn't used internally.
+    case unexpected
+}
+
+/// A UDP server that can be used for signalling between the UI tests jig and the app.
+/// The server should be instantiated on the UI tests side.
+class UITestSignalServer: UITestSignalProtocol {
```
Comment (stefanceriu) [inline ElementX/Sources/UITests/UITestSignalling.swift:48]: We should probably call this `UITests` to match the other classes.
I was also under the impression that we'll go with classes nested withing it i.e. `UITestsSignalling.Server` and `UITestsSignalling.Client`

File: ElementX/Sources/UITests/UITestSignalling.swift [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    let listener: NWListener
```
Comment (stefanceriu) [inline ElementX/Sources/UITests/UITestSignalling.swift:49]: This one can be private

File: ElementX/Sources/UITests/UITestSignalling.swift [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        connection.cancel()
+        if let nextMessageContinuation {
+            nextMessageContinuation.resume(throwing: UITestSignalError.cancelled)
+            self.nextMessageContinuation = nil
+        }
+    }
+}
+
+/// A shared implementation for sending/receiving signals between a client and a server.
+protocol UITestSignalProtocol: AnyObject {
```
Comment (stefanceriu) [inline ElementX/Sources/UITests/UITestSignalling.swift:140]:
```suggestion
protocol UITestsSignallingProtocol: AnyObject {
```

File: UITests/Sources/RoomScreenUITests.swift [hunk 5]
```diff
@@ -47,19 +47,123 @@ class RoomScreenUITests: XCTestCase {
...
+        
+        // Then the UI should still remain unchanged.
+        app.assertScreenshot(.roomLayoutMiddle, step: 0)
+        
+        // When the keyboard appears for the message composer.
+        try await tapMessageComposer(in: app)
+        
+        // Then the timeline scroll offset should remain unchanged.
+        app.assertScreenshot(.roomLayoutMiddle, step: 1)
+        signal.disconnect()
```
Comment (stefanceriu) [inline UITests/Sources/RoomScreenUITests.swift:110]: Would probably be cleaner to use a `defer` for these

File: ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift [hunk 6]
```diff
@@ -20,52 +20,26 @@ import Foundation
...
-        
-        Task {
-            try await Task.sleep(for: incomingDelay)
-            timelineItems.append(incomingItem)
-            callbacks.send(.updatedTimelineItems)
-            
-            if !self.incomingItems.isEmpty {
-                simulateIncomingItems()
-            }
+    private let signal = UITestSignalClient()
```
Comment (stefanceriu) [inline ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift:32]: Do we really want to create this from the get go?

File: ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift [hunk 7]
```diff
@@ -87,4 +61,47 @@ class MockRoomTimelineController: RoomTimelineControllerProtocol {
     func debugDescriptionFor(_ itemId: String) -> String {
         "Mock debug description"
     }
+    
+    // MARK: - UI Test signalling
+    
+    /// Allows the simulation of server responses by listening for signals from UI tests.
+    private func enableServerSimulation() async throws {
```
Comment (stefanceriu) [inline ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift:68]: Not sure "server simulation" is the best name for what we're doing here

File: ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift [hunk 8]
```diff
@@ -87,4 +61,47 @@ class MockRoomTimelineController: RoomTimelineControllerProtocol {
...
+            switch try await signal.receive() {
+            case .paginate:
+                try await self.simulateBackPagination()
+            case .incomingMessage:
+                try await self.simulateIncomingItem()
+            default:
+                break
+            }
+            
+            // Keep listening TODO: Try an async stream here?
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift:80]: Yes please! And if that doesn't work move this whole thing to a normal loop

**Review by stefanceriu — CHANGES_REQUESTED** [review]
This looks and works great, nicely done! Left some comments inline

**Comment by stefanceriu** [comment]
Note, I've manually disabled CI UI tests until this PR is merged as they keep failing every night

File: ElementX/Sources/UITests/UITestSignalling.swift (hunk 2 above)
Comment (pixlwave) [reply ElementX/Sources/UITests/UITestSignalling.swift:48]: Oh yeah, I looked at `UITestScreenIdentifier` and didn't notice the others are plural. Makes sense 👍

File: ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift (hunk 6 above)
Comment (pixlwave) [reply ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift:32]: Probably not 🙃

File: ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (pixlwave) [reply ElementX/Sources/Services/Timeline/TimelineController/MockRoomTimelineController.swift]: I went with the normal loop in the end.

File: ElementX/Sources/Screens/RoomScreen/View/Style/TimelineItemBubbledStylerView.swift [hunk 9]
```diff
@@ -42,6 +42,7 @@ struct TimelineItemBubbledStylerView<Content: View>: View {
                     styledContentWithReactions
                 }
                 .padding(.horizontal, 16.0)
+                .padding(timelineItem.isOutgoing ? .leading : .trailing, 40) // Extra padding to differentiate alignment.
```
Comment (pixlwave) [inline ElementX/Sources/Screens/RoomScreen/View/Style/TimelineItemBubbledStylerView.swift:45]: I don't think this change should have happened - it more closely matches Figma this way.

File: UITests/Sources/Application.swift [hunk 10]
```diff
@@ -41,13 +41,21 @@ extension XCUIApplication {
 
     /// Assert screenshot for a screen with the given identifier. Does not fail if a screenshot is newly created.
     /// - Parameter identifier: Identifier of the UI test screen
-    func assertScreenshot(_ identifier: UITestScreenIdentifier, step: Int? = nil) {
+    /// - Parameter step: An optional integer that can be used to take multiple snapshots per test identifier.
+    /// - Parameter insets: Optional insets with which to crop the image by.
+    func assertScreenshot(_ identifier: UITestScreenIdentifier, step: Int? = nil, insets: UIEdgeInsets? = nil) {
```
Comment (pixlwave) [inline UITests/Sources/Application.swift:46]: I've added the option to crop the snapshot to ignore for example the navigation bar which may change in opacity depending on whether a scrollview is at the top or has content under the bar.

**Pull request merged**